Support for response_type=token

justinabrahms opened the pull request:
# Overview

This change adds support for implicit grant oauth status. This change is necessary to support browser-only OAuth clients, which we'll be making use of as part of some CCX work at MIT's ODL. This does not remove functionality, but rather provides another avenue for OAuth2 clients.
# Manual testing

You should be able to issue a request to the following URL (note that you'll need to update the client_id and redirect_uri to match what's on your local system). The result should redirect you to localhost:9001 with a URL fragment containing the access token.

http://localhost:8000/oauth2/authorize/?client_id=015c91a76033cac3040d&redirect_uri=localhost:9001&response_type=token

Example result: http://localhost:9001/#access_token=a16051276d052616bcb2f87e71974aaedab9bdfa&token_type=Bearer&expires_in=2591999&scope=
# Background

This is mostly code from https://github.com/caffeinehit/django-oauth2-provider/pull/59 with the exception of a few tests that I wrote.

It relies on #13.

I've not communicated with the greater edX team about this work in particular and there is no backing JIRA ticket.

kusamau committed:
Implements the 'Implicit Grant'.
A typical call looks like
/oauth2/authorize/?client_id=an_existing_client_id=token&redirect_uri=the_client_redirect_uri

Implements the 4.2.2 "Access Token Response" specifications

kusamau committed:
Implements the 'Implicit Grant'.
A typical call looks like
/oauth2/authorize/?client_id=an_existing_client_id=token&redirect_uri=the_client_redirect_uri

kusamau committed:
Implements the 4.2.2 "Access Token Response" specifications

kusamau committed:
Merge branch 'master' of https://github.com/kusamau/django-oauth2-provider

kusamau committed:
A 'response_type' == 'token' must return a json object leaving the
redirect_uri to the client

Removes unused parameter

kusamau committed:
A 'response_type' == 'token' must return a json object leaving the
redirect_uri to the client

kusamau committed:
Removes unused parameter

kusamau committed:
no data parameter required

A 'response_type' == 'token' must return a json object leaving the
redirect_uri to the client

Removes unused parameter

kusamau committed:
Merge branch 'master' of https://github.com/kusamau/django-oauth2-provider

kusamau committed:
Revert "Implements the 'Implicit Grant'. A typical call looks like /oauth2/authorize/?client_id=an_existing_client_id=token&redirect_uri=the_client_redirect_uri"

This reverts commit 728322173d3718e7cbbd8e216fbc22500d737b61.

kusamau committed:
Revert "Revert "Implements the 'Implicit Grant'. A typical call looks like /oauth2/authorize/?client_id=an_existing_client_id=token&redirect_uri=the_client_redirect_uri""

This reverts commit b63b27be60d6851a862051ed5b278642d57b721a.

kusamau committed:
Revert the implicit grant with (some corrections)

kusamau committed:
rollback

kusamau committed:
Missing params

kusamau committed:
missing param2

kusamau committed:
missing import

justinabrahms committed:
Added support for implicit grant access type.

This is necessary to write browser-only oauth2 clients.

Original code (sans-test) from 'kusamau/master'

openedx-webhooks commented:
Thanks for the pull request, @justinabrahms! I've created [OSPR-942](https://openedx.atlassian.net/browse/OSPR-942) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).
We can't start reviewing your pull request until you've   added yourself to the [AUTHORS](https://github.com/edx/django-oauth2-provider/blob/master/AUTHORS) file . Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst) file for more information.

sarina commented:
@justinabrahms Hi, I'm sorry, I'm totally unfamiliar with this repo. I have no idea what you're doing or what team would look at this. Could you please provide way more detail when you open PRs? See http://edx-developer-guide.readthedocs.org/en/latest/process/cover-letter.html

justinabrahms commented:
Sorry, Sarina. I'm with ODL and pretty new. I'll update it with this
tomorrow morning first thing.

On Monday, November 9, 2015, Sarina Canelake notifications@github.com
wrote:

> @justinabrahms https://github.com/justinabrahms Hi, I'm sorry, I'm
> totally unfamiliar with this repo. I have no idea what you're doing or what
> team would look at this. Could you please provide way more detail when you
> open PRs? See
> http://edx-developer-guide.readthedocs.org/en/latest/process/cover-letter.html
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/edx/django-oauth2-provider/pull/14#issuecomment-155230071
> .

## 

Sent via mobile phone.

sarina commented:
Hey Justin, no worries. Before you open more PRs, it's probably a good idea
to talk with your colleagues so they can share tips. Also, a trip through
our contributing docs would be very helpful! (I know they're a bit long and
confusing - we should definitely streamline them... someday... sigh.)

On Mon, Nov 9, 2015 at 6:19 PM, Justin Abrahms notifications@github.com
wrote:

> Sorry, Sarina. I'm with ODL and pretty new. I'll update it with this
> tomorrow morning first thing.
> 
> On Monday, November 9, 2015, Sarina Canelake notifications@github.com
> wrote:
> 
> > @justinabrahms https://github.com/justinabrahms Hi, I'm sorry, I'm
> > totally unfamiliar with this repo. I have no idea what you're doing or
> > what
> > team would look at this. Could you please provide way more detail when
> > you
> > open PRs? See
> > 
> > http://edx-developer-guide.readthedocs.org/en/latest/process/cover-letter.html
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > <
> > https://github.com/edx/django-oauth2-provider/pull/14#issuecomment-155230071
> > 
> > .
> 
> ## 
> 
> Sent via mobile phone.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/edx/django-oauth2-provider/pull/14#issuecomment-155230317
> .

clintonb commented:
@nasthagiri @mattdrayer Can one of you also review this change, or recommend someone else, please?

justinabrahms commented:
I've found a bug wherein users are not properly prompted with the "do you grant access" interstitial. I'm moving this back to work-in-progress status until this is solved. You can hold off on reviews until this is done.

nasthagiri commented:
@jimabramson FYI.

jimabramson commented:
@justinabrahms in case it's helpful, our oauth2 provider adds the notion of a "trusted client" (see https://github.com/edx/edx-oauth2-provider/blob/master/oauth2_provider/models.py) which is something we use to deliberately disable the interstitial (for more seamless SSO across internal services).  Not sure where you're running into the issue you're having, but just mentioning it since in some typical edx.org configurations you'll find that disabled by design.

justinabrahms committed:
Added support for grant_type=token

Differences from the work of kusamau include:

1. Additional tests.
2. Verifies AccessToken expiration & scoping rules.
3. Ensures compatibility with SINGLE_ACCESS_TOKEN setting.
4. Allows user to cancel their authorization request.
5. General code cleanup for clarity.
6. Move toward use of .format over % for interpolation.
7. Fixed a bug causing intermittant test failures.

justinabrahms commented:
Thank you, @jimabramson. I haven't configured this test client to be a trusted client. This was a bug in indentation, so it seems. I've written tests to validate that this does in fact fix it. The difference, if you want to manually test it, is in the previous code, you'd visit the URLs in the description and it'd immediately give you a token. Now, it will give you the interstitial, and upon acceptance.. the token.

This is, once again, ready for review.

jimabramson commented on provider/oauth2/tests.py:
```diff
@@ -126,6 +126,20 @@ def test_authorization_requires_supported_response_type(self):
+    def test_token_authorization_redirects_to_correct_uri(self):
+        self.login()
+
+        response1 = self.client.get(self.auth_url() + '?client_id=%s&response_type=token' % self.get_client().client_id, follow=True)
```
`.format` is preferred to string interpolation.   also, any reason not to use the `data` kwarg instead of doing string concatenation?

justinabrahms replied:
Is `.format` preferred over keeping consistent style with the repository itself?

The reason for not using data was because other API calls (see method just above this one) didn't.

jimabramson replied:
@justinabrahms ah, you updated this just as I was going to respond :-)

In general we try to stick with a mindset of improving/updating code style where possible when adding/changing stuff.  That does sometimes conflict with the notion of maintaining internal consistency as you've pointed out, and also, this being a forked repo, it's even more of a special case.  So it wouldn't have been a dealbreaker, but appreciate your making the changes.

jimabramson commented on provider/oauth2/tests.py:
```diff
@@ -579,7 +593,7 @@ def test_clear_expired(self):
 
         # verify that Grant with code exists
-        code = urlparse.parse_qs(location)['code'][0]
+        code = urlparse.parse_qs(location.split('?')[1])['code'][0]
```
nice catch here.  though it's slightly less concise, `urlparse.urlparse(location).query` would be clearer here.

jimabramson commented on provider/views.py:
```diff
@@ -261,12 +263,22 @@ def handle(self, request, post_data=None):
+        
+        if data.get('response_type', None):
+            if not authorization_form.is_bound \
+                or not authorization_form.is_valid():
```
no need to break this line.

optional: not (x and y) would read a little better.

jimabramson commented on provider/views.py:
```diff
@@ -261,12 +263,22 @@ def handle(self, request, post_data=None):
+                #uses request.user as the urls.py already required login_required
+                atm = AccessTokenModel.objects.create(user=request.user,
+                                                client=client,
+                                                scope=data.get('scope'))
```
When the authorization request omits the 'scope' parameter, this causes an empty `scope=` to be returned in the response, which is technically incorrect in terms of [the rfc](https://tools.ietf.org/html/rfc6749#section-3.3) because it implies that an "empty" scope has been granted.  I think only including 'scope' when it was specified in the request would be a reasonable fix.

jimabramson commented on provider/views.py:
```diff
@@ -489,13 +501,23 @@ def access_token_response_data(self, access_token):
+            if len(data.get('state', '')) > 0:
+                response_data['state'] = data.get('state')
+            path = "%s#%s" % (basepath,
+                              urlencode(response_data))
```
Please use string format instead of interpolation.

jimabramson commented on provider/views.py:
```diff
@@ -261,12 +263,22 @@ def handle(self, request, post_data=None):
+
+            if data.get('response_type') == 'token':
+                #uses request.user as the urls.py already required login_required
+                atm = AccessTokenModel.objects.create(user=request.user,
```
the base provider views check a `SINGLE_ACCESS_TOKEN` constant and decide whether or not to issue a new token or reuse an existing one (if exists).  Would be ideal to maintain consistency with that behavior here.

jimabramson commented:
@nasthagiri @mattdrayer @clintonb were any of you planning on reviewing this as well?

justinabrahms commented:
GitHub isn't being helpful w/ their UI. 

Jim: I still need to address your `SINGLE_ACCESS_TOKEN` feedback.

justinabrahms commented:
`SINGLE_ACCESS_TOKEN` feedback has been addressed.

jimabramson commented on provider/views.py:
```diff
@@ -262,11 +265,48 @@ def handle(self, request, post_data=None):
+                    "error": "access_denied",
+                    "error_description": "The user denied the scope request.",
+                    "state": data.get('state'),
+                }
```
nice catch on this.

jimabramson commented on provider/views.py:
```diff
@@ -262,11 +265,48 @@ def handle(self, request, post_data=None):
+                atm = None
+                if constants.SINGLE_ACCESS_TOKEN:
+                    try:
+                        atm = AccessTokenModel.objects.get(user=request.user)
```
`get` needs to be filtered by user, client, and scope, else we risk leaking tokens meant for other clients/scopes.  (it's correct here: https://github.com/justinabrahms/django-oauth2-provider/blob/master/provider/oauth2/views.py#L95)

jimabramson replied:
oops, master has a bug that this branch does not
https://github.com/justinabrahms/django-oauth2-provider/blob/grant_type_token/provider/oauth2/views.py#L106

the expiration date has to be checked as well.

jimabramson commented:
@justinabrahms after fixing the `get()` call (and squashing up commits), I am ready to merge this PR.

I have plans to build something on top of your work once it's merged, so please let me know if there's anything I can do to help bring this in for a landing.  Thanks!

justinabrahms commented:
Thanks Jim. I'll have this cleaned up by eod.

justinabrahms commented:
This should cover the case you'd like. After review, let me know and I'll squash.

justinabrahms commented:
You should know (for your own use case) that before this becomes 'super officially supported' there will need to be some templating work done. It's a rather unstyled acceptance page.

jimabramson commented:
@justinabrahms the interstitial was already rough and not affected by your work, if i understand correctly.  We don't use that for edx.org anyway AFAIK, but thank you for calling it out.

jimabramson commented:
@justinabrahms LGTM.  squashing time!   My suggestion is to squash your own commits but maybe leave the ones from https://github.com/caffeinehit/django-oauth2-provider/pull/59 as they are.  It'll look messy but I think it might be helpful to others when considering if/how to employ this fork.   (@clintonb speak up if you have any strong feelings about that?)

justinabrahms commented:
Squashed!

clintonb commented:
I'm all for giving credit where it's due.

jimabramson commented:
@justinabrahms thanks for the contribution!

justinabrahms commented:
Happy to help. Any pointers on who I talk with to discuss the templating
side of this?

On Friday, November 20, 2015, Jim Abramson notifications@github.com wrote:

> @justinabrahms https://github.com/justinabrahms thanks for the
> contribution!
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/edx/django-oauth2-provider/pull/14#issuecomment-158517853
> .

## 

Sent via mobile phone.

clintonb commented:
Our UI/UX resources are spread thin. It might be quicker to propose your own design, and use it as a baseline.
